src/linear_algebra/finsupp.lean

```diff
@@ -141,247 +101,252 @@ begin
+
+theorem supported_Inter {ι : Type*} (s : ι → set α) :
+  supported β γ (⋂ i, s i) = ⨅ i, supported β γ (s i) :=
 begin
-  have eq : _inst_6 = (λa b : β, classical.prop_decidable (a = b)) := subsingleton.elim _ _,
-  unfreezeI, subst eq,
-  refine (finsupp.restrict_dom_equiv_finsupp α α s)
+  refine le_antisymm (le_infi $ λ i, supported_mono $ set.Inter_subset _ _) _,
+  simp [le_def, infi_coe, set.subset_def],
+  exact λ l, set.subset_Inter
 end
 
-end lc
-
-section vector_space
-universes u v
-variables {α : Type u} {β γ : Type v}
-variables [discrete_field α]
-variables [add_comm_group β] [vector_space α β]
-variables [add_comm_group γ] [vector_space α γ]
+set_option class.instance_max_depth 37
+def supported_equiv_finsupp (s : set α) [decidable_pred (λ (x : α), x ∈ s)] :
+  (supported β γ s) ≃ₗ[γ] (s →₀ β) :=
+(restrict_support_equiv s).to_linear_equiv
+begin
+  show is_linear_map γ ((lsubtype_domain s : (α →₀ β) →ₗ[γ] (s →₀ β)).comp
+      (submodule.subtype (supported β γ s))),
+  exact linear_map.is_linear _
+end
+set_option class.instance_max_depth 32
 
-open vector_space
+def lsum (f : α → γ →ₗ[γ] β) : α →₀ γ →ₗ[γ] β :=
+⟨λ d, d.sum (λ i, f i),
+  assume d₁ d₂, by simp [sum_add_index],
+  assume a d, by simp [sum_smul_index, smul_sum, -smul_eq_mul, smul_eq_mul.symm]⟩
 
-set_option class.instance_max_depth 100
+@[simp] theorem lsum_apply (f : α → γ →ₗ[γ] β) (l : α →₀ γ) :
+  (finsupp.lsum f : α →₀ γ →ₗ β) l = l.sum (λ b, f b) := rfl
 
-lemma equiv_of_dim_eq_dim (h : dim α β = dim α γ) : nonempty (β ≃ₗ[α] γ) :=
-begin
-  rcases exists_is_basis α β with ⟨b, hb⟩,
-  rcases exists_is_basis α γ with ⟨c, hc⟩,
-  rw [← hb.mk_eq_dim, ← hc.mk_eq_dim] at h,
-  rcases quotient.exact h with ⟨e⟩,
-  exact ⟨ (module_equiv_lc hb).trans (linear_equiv.trans (lc.congr α b c e) (module_equiv_lc hc).symm) ⟩
-end
+section lmap_domain
+variables {α' : Type*} [decidable_eq α'] {α'' : Type*} [decidable_eq α''] (β γ)
 
-lemma eq_bot_iff_dim_eq_zero (p : submodule α β) (h : dim α p = 0) : p = ⊥ :=
-begin
-  have : dim α p = dim α (⊥ : submodule α β) := by rwa [dim_bot],
-  rcases equiv_of_dim_eq_dim this with ⟨e⟩,
-  exact e.eq_bot_of_equiv _
-end
+def lmap_domain (f : α → α') : (α →₀ β) →ₗ[γ] (α' →₀ β) :=
+⟨map_domain f, assume a b, map_domain_add, map_domain_smul⟩
 
-lemma injective_of_surjective (f : β →ₗ[α] γ)
-  (hβ : dim α β < cardinal.omega) (heq : dim α γ = dim α β) (hf : f.range = ⊤) : f.ker = ⊥ :=
-have hk : dim α f.ker < cardinal.omega := lt_of_le_of_lt (dim_submodule_le _) hβ,
-begin
-  rcases cardinal.lt_omega.1 hβ with ⟨d₁, eq₁⟩,
-  rcases cardinal.lt_omega.1 hk with ⟨d₂, eq₂⟩,
-  have : 0 = d₂,
-  { have := dim_eq_surjective f (linear_map.range_eq_top.1 hf),
-    rw [heq, eq₁, eq₂, ← nat.cast_add, cardinal.nat_cast_inj] at this,
-    exact nat.add_left_cancel this },
-  refine eq_bot_iff_dim_eq_zero _ _,
-  rw [eq₂, ← this, nat.cast_zero]
-end
+@[simp] theorem map_apply (f : α → α') (l : α →₀ β) :
```

Comment:
this should be `lmap_domain_apply`

Reply:
Renamed to `lmap_domain_apply` (and similarly for the other lemmas).